Render error tx844

sohammk08 opened the pull request:
## High Level Overview of Change

Made changes in translations.json to exclude _connection.server.publicKey_ from **server_ledgers_hint**

### Context of Change
This change was made because the _connection.server.publicKey_ in **server_ledgers_hint** in translations.json wasn't able to function properly. Upon closer inspection, it was found that the reason for this occurrence was the _publicKey_ field being empty. Removed the _connection.server.publicKey_ part and restructured sentences accordingly.

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->


- [x] Translation Updates

### TypeScript/Hooks Update

<!--
In an effort to modernize the codebase, you should convert the files that you work with to React Hooks and TypeScript.
If this is not possible (e.g. it's too many changes, touching too many files, etc.) please explain why here.
-->

- [ ] Updated files to React Hooks
- [ ] Updated files to TypeScript

## Before / After
Before:
```json
"server_ledgers_hint": "This node {{connection.server.publicKey}} only contains ledgers {{connection.ledger.validated}}",
```
After:
```json
"server_ledgers_hint": "This node only contains ledgers {{connection.ledger.validated}}",
```

## Test Plan

<!--
Please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

mvadari commented on public/locales/en-US/translations.json:
```diff
@@ -200,7 +200,7 @@
   "invalid_transaction_hash": "The transaction hash is invalid",
   "ledger_not_found": "Ledger not Found",
   "check_ledger_id": "Please check your ledger id",
-  "server_ledgers_hint": "This node ({{connection.server.publicKey, truncate(length: 10)}}) only contains ledgers {{connection.ledger.validated}}",
```
I think we still want the pubkey to be printed if it's available, with some sort of "This Clio node only..." alternate message

sohammk08 replied:
So, you're saying that the **clio-node-part** still needs to be present in the code. What if I change the code such that whenever the **pubKey** is not empty, it gets displayed. But when it is empty, print it as "This clio node only..."

Would that be viable?

mvadari replied:
Yes, something like that sounds good.

sohammk08 commented:
@ckniffen @pdp2121 @mvadari 
I finally have an update. I managed to get the publicKey part working.

However, I'm still not sure about the part where we display **server_ledgers_hint** if publicKey does not exist and **server_ledgers_hint_with_key** if it does.

Here is where I currently am, for reference:

![Capture](https://github.com/ripple/explorer/assets/87537892/cffebaa7-8781-4e91-a17a-986cff2bf728)

pdp2121 commented:
> @ckniffen @pdp2121 @mvadari I finally have an update. I managed to get the publicKey part working.
> 
> However, I'm still not sure about the part where we display **server_ledgers_hint** if publicKey does not exist and **server_ledgers_hint_with_key** if it does.
> 
> Here is where I currently am, for reference:
> 
> ![Capture](https://private-user-images.githubusercontent.com/87537892/311450792-cffebaa7-8781-4e91-a17a-986cff2bf728.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNzY0NDcsIm5iZiI6MTcxMjA3NjE0NywicGF0aCI6Ii84NzUzNzg5Mi8zMTE0NTA3OTItY2ZmZWJhYTctODc4MS00ZTkxLWExN2EtOTg2Y2ZmMmJmNzI4LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE2NDIyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyMjI5NDU3MDdmN2EyMmI4ODk2MmEzY2M5ZTVmMGVhMzdjMmM1MGNmMzc3YTZkYTZlNTE5NDg4ZDZlNmI3NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OuJM_HTllInSFFmqowSQGsVTMzHvpuJ7gqeVY3u-p94)

@sohammk08  You could use ternary operator to condition on whether the publicKey exists and display different messages based on that.

sohammk08 commented:
@pdp2121 @mvadari @ckniffen I have solved the error and now, it works properly. Just to make sure, I have also tested to see if it works, and it does. I am giving below the method I used to test. I will be making a commit containing the changes soon after writing this comment. Below is how I tested it to make sure if it works or not. Take a look:

Original file:
```typescript
  const notFound = title.includes('not_found')
  const hintMsg = hints.map((hint) => (
    <div className="hint" key={hint}>
      {t(hint as any, values)}
    </div>
  ))
  const derivedWarning = warning ?? (notFound && t('not_found'))

  return (
    <div className="no-match">
      <Helmet title={t(title as any)} />
      {isError && <div className="uh-oh">{t('uh_oh')}</div>}
      <div className="title">{t(title as any, values)}</div>
      {hintMsg}
      {(derivedWarning || isError) && (
        <div className="warning">
          <InfoIcon title={derivedWarning} />
          &nbsp;
          <span>{derivedWarning}</span>
        </div>
      )}
    </div>
  )
}

export default NoMatch
```

Changed file:
``` typescript
  const notFound = title.includes('not_found')

  // Determine which hint message to display based on the presence of publicKey
  const hintMsg = values.connection?.server?.publicKey
    ? t('server_ledgers_hint_with_key', values)
    : t('server_ledgers_hint', values)

  const derivedWarning = warning ?? (notFound && t('not_found'))

  return (
    <div className="no-match">
      <Helmet title={t(title as any)} />
      {isError && <div className="uh-oh">{t('uh_oh')}</div>}
      <div className="title">{t(title as any, values)}</div>
      {hintMsg && <div className="hint">{hintMsg}</div>}
      {(derivedWarning || isError) && (
        <div className="warning">
          <InfoIcon title={derivedWarning} />
          &nbsp;
          <span>{derivedWarning}</span>
        </div>
      )}
    </div>
  )
}

export default NoMatch
```

I also made changes to translations.json as such:
``` json
  "server_ledgers_hint": "This node only contains ledgers {{connection.ledger.validated}}",
  "server_ledgers_hint_with_key": "This node ({{connection.server.publicKey, truncate(length: 10)}}) only contains ledgers {{connection.ledger.validated}}",
```

Also, here is how I tested whether the change works or not;
First, I set the publicKey to null for debugging purposes. Then, I changed the hintMsg line. Here is how it looks in code:
```typescript
  const notFound = title.includes('not_found')
  const publicKey = null

  // Determine which hint message to display based on the presence of publicKey
  const hintMsg = publicKey
    ? t('server_ledgers_hint_with_key', values)
    : t('server_ledgers_hint', values)

  const derivedWarning = warning ?? (notFound && t('not_found'))
```

I checked the logic after adding the above debugging code as said before. Also, please let me know if I should try and implement a test in NoMatch.test.tsx for the same.

Have a good day :)

pdp2121 commented:
> @pdp2121 @mvadari @ckniffen I have solved the error and now, it works properly. Just to make sure, I have also tested to see if it works, and it does. I am giving below the method I used to test. I will be making a commit containing the changes soon after writing this comment. Below is how I tested it to make sure if it works or not. Take a look:
> 
> Original file:
> 
> ```ts
>   const notFound = title.includes('not_found')
>   const hintMsg = hints.map((hint) => (
>     <div className="hint" key={hint}>
>       {t(hint as any, values)}
>     </div>
>   ))
>   const derivedWarning = warning ?? (notFound && t('not_found'))
> 
>   return (
>     <div className="no-match">
>       <Helmet title={t(title as any)} />
>       {isError && <div className="uh-oh">{t('uh_oh')}</div>}
>       <div className="title">{t(title as any, values)}</div>
>       {hintMsg}
>       {(derivedWarning || isError) && (
>         <div className="warning">
>           <InfoIcon title={derivedWarning} />
>           &nbsp;
>           <span>{derivedWarning}</span>
>         </div>
>       )}
>     </div>
>   )
> }
> 
> export default NoMatch
> ```
> 
> Changed file:
> 
> ```ts
>   const notFound = title.includes('not_found')
> 
>   // Determine which hint message to display based on the presence of publicKey
>   const hintMsg = values.connection?.server?.publicKey
>     ? t('server_ledgers_hint_with_key', values)
>     : t('server_ledgers_hint', values)
> 
>   const derivedWarning = warning ?? (notFound && t('not_found'))
> 
>   return (
>     <div className="no-match">
>       <Helmet title={t(title as any)} />
>       {isError && <div className="uh-oh">{t('uh_oh')}</div>}
>       <div className="title">{t(title as any, values)}</div>
>       {hintMsg && <div className="hint">{hintMsg}</div>}
>       {(derivedWarning || isError) && (
>         <div className="warning">
>           <InfoIcon title={derivedWarning} />
>           &nbsp;
>           <span>{derivedWarning}</span>
>         </div>
>       )}
>     </div>
>   )
> }
> 
> export default NoMatch
> ```
> 
> I also made changes to translations.json as such:
> 
> ```json
>   "server_ledgers_hint": "This node only contains ledgers {{connection.ledger.validated}}",
>   "server_ledgers_hint_with_key": "This node ({{connection.server.publicKey, truncate(length: 10)}}) only contains ledgers {{connection.ledger.validated}}",
> ```
> 
> Also, here is how I tested whether the change works or not; First, I set the publicKey to null for debugging purposes. Then, I changed the hintMsg line. Here is how it looks in code:
> 
> ```ts
>   const notFound = title.includes('not_found')
>   const publicKey = null
> 
>   // Determine which hint message to display based on the presence of publicKey
>   const hintMsg = publicKey
>     ? t('server_ledgers_hint_with_key', values)
>     : t('server_ledgers_hint', values)
> 
>   const derivedWarning = warning ?? (notFound && t('not_found'))
> ```
> 
> I checked the logic after adding the above debugging code as said before. Please let me know if I should try and implement a test in NoMatch.test.tsx for the same. Also, as @pdp2121 suggested, I set the server_ledgers_hint_with_key for other language translations as _null_
> 
> Have a good day :)

Hi @sohammk08, it would be nice if you could add a test inside NoMatch.test.tsx. Thanks for your contribution!

pdp2121 commented on src/containers/NoMatch/index.tsx:
```diff
@@ -40,19 +40,20 @@ const NoMatch = ({
       {isError && <div className="uh-oh">{t('uh_oh')}</div>}
       <div className="title">{t(title as any, values)}</div>
-      {hintMsg}
+      {hintMsg && <div className="hint">{hintMsg}</div>}
```
I'm not sure why we need to change here. Would `hintMsg` ever be `null`? Also, are we not using `hints` array here at all?

sohammk08 replied:
Correct, at present, the _hints_ array isn't directly used within the JSX rendering logic of the component.
Instead, it's utilized within the _useEffect_ hook to concatenate its values into the _description_ parameter of the _track_ function.

sohammk08 commented:
> Tests are failing since we are using node 18 instead.

Yes. I checked the staging branch of the ripple/staging and all tests are passing. I ran _npm run test_ and almost 11 tests are failing. Hopefully, I will be committing the revised code such that it doesn't fail any tests and.
Thanks for your input.

Good Day.
